[Jetpack Content Migration Flow] Refactor login resolver as orchestrator

mkevins opened the pull request:
# Description

This PR continues from #17398 by refactoring the login resolver as an orchestrator. With the unified entity provider, the entity-specific logic is extracted to a helper class for the provider-side. The resolver-side is simplified to a single inlined query to obtain the data, leaving the consumption of that data to the orchestrator.

This begins the process and serves as a first step of unifying the various providers, and reducing the boilerplate needed to retrieve specified data from the provider. It also renames the existing shared login resolver class as an orchestrator to better describe its responsibility.

To test:

Follow the testing steps from: https://github.com/wordpress-mobile/WordPress-Android/pull/17398, the behavior should not change.

## Regression Notes
1. Potential unintended areas of impact
No. Existing implementations are changed, but are behind a feature flag.

9. What I did to test those areas of impact (or what existing automated tests I relied on)
Existing implementations have automated tests, and manual testing was also done to ensure things are still working as desired.

10. What automated tests I added (or what prevented me from doing so)
Unit tests were added for these implementations.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mkevins committed:
Refactor shared login resolver as orchestrator

This commit begins the process of unifying the local content providers
and refactors the existing logic in the shared login resolver as an
orchestrator.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/localcontentmigration/LocalAccessTokenProviderHelper.kt:
```diff
@@ -0,0 +1,11 @@
+class LocalAccessTokenProviderHelper @Inject constructor(
+    private val accountStore: AccountStore,
+): LocalDataProviderHelper {
+    override fun getData(localSiteId: Int?, localEntityId: Int?) = AccessTokenData(token = accountStore.accessToken)
```
Would it make sense to unit test the return here, and also if `accountStore.accessToken` is being called?

mkevins replied:
I'm on the fence about whether this could be tested.. but if we do add a unit test for this helper, maybe we test that `getData` returns the correct `AccessTokenData` (including that it has the correct token, from a mock), but perhaps not that it invokes a particular getter. I.e. state verification instead of behavior verification. Wdyt?

mkevins replied:
I've added a test here: https://github.com/wordpress-mobile/WordPress-Android/pull/17399/commits/2b0607ee5d20ef5f16917a523bf9f4908cf08ecc.

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/localcontentmigration/LocalMigrationContentResolver.kt:
```diff
@@ -30,12 +30,12 @@ private fun ContentResolver.query(
-    private val contextProvider: ContextProvider,
-    private val wordPressPublicData: WordPressPublicData,
-    private val queryResult: QueryResult,
+    val contextProvider: ContextProvider,
```
I just wanted to confirm if we really need to remove `private` modifier from these parameters.

mkevins replied:
Sadly, due to the way these inline calls are chained, it seems we need to increase the visibility. I'd be happy if there was a way to keep them private though, if you have a suggestion :thinking:

mkevins replied:
So, it turns out there is a way :smile: We can use an annotation to make these less visible in Kotlin (though the names are still visible in the JVM). This is an improvement though :+1: I've addressed this in this commit: https://github.com/wordpress-mobile/WordPress-Android/pull/17399/commits/4db652422bcb5847cc02deb10ebc21dcc5bad907

RenanLukas commented on WordPress/src/main/java/org/wordpress/android/sharedlogin/resolver/LocalMigrationOrchestrator.kt:
```diff
@@ -47,9 +41,10 @@ class SharedLoginResolver @Inject constructor(
-        if (accessTokenCursor != null) {
-            val accessToken = queryResult.getValue<String>(accessTokenCursor) ?: ""
+        val (accessToken) = localMigrationContentResolver.getDataForEntityType<AccessTokenData>(AccessToken)
+//        val accessTokenCursor = getAccessTokenCursor()
```
Would it make sense to remove the commented block before merging? Or we want to handle it in another PR?

mkevins replied:
Good question.. I left these commented lines since I didn't want to delete the existing error information. The error handling may be best to move to a separate PR, but I think that the existing work on errors will be reusable in the unified approach. So, this was like a "reminder" of a loose end that needs to be re-included somewhere. That said, I'm happy to remove these commented lines, and the existing names can be "recycled" / repurposed from the commit history :smile: .

mkevins replied:
Handled with: https://github.com/wordpress-mobile/WordPress-Android/pull/17399/commits/1eb46108f518d578bdc4161d09de33766e11f14a

RenanLukas commented on WordPress/src/test/java/org/wordpress/android/sharedlogin/resolver/SharedLoginResolverTest.kt:
```diff
@@ -18,19 +18,21 @@ import org.wordpress.android.fluxc.Dispatcher
 import org.wordpress.android.util.AccountActionBuilderWrapper
 import org.wordpress.android.viewmodel.ContextProvider
 
+// TODO: adapt these tests to the unified provider / orchestrator approach
```
Should we handle this before merging or in a different PR?

mkevins replied:
Good question.. maybe we could try tackling these in this PR? The rationale being that our tests can help identify issues worth addressing in the implementation, and once it is more solid, later PRs based on this will have a stronger foundation. Wdyt?

mkevins replied:
I had a second thought about this, and maybe this makes more sense after the other resolver / provider pairs are unified. My thinking is, by then, the error classes which dictate flow control will be more clearly defined, so we'll have a better idea of what to test for the orchestrator. Wdyt?

RenanLukas reviewed (commented):
Hey @mkevins , thanks for the PR!
I've left a few comments. Please let me know if they make sense.

mkevins committed:
Merge branch 'try/local-migration-content-provider' into try/local-migration-content-provider--refactor-login-resolver-as-orchestrator

mkevins committed:
Merge branch 'try/local-migration-content-provider' into try/local-migration-content-provider--refactor-login-resolver-as-orchestrator

mkevins committed:
Merge branch 'try/local-migration-content-provider' into try/local-migration-content-provider--refactor-login-resolver-as-orchestrator

mkevins committed:
Merge branch 'trunk' into try/local-migration-content-provider--refactor-login-resolver-as-orchestrator

mkevins committed:
Fix indentation and remove commented code

mkevins committed:
Suppress TODO warning

mkevins committed:
Use @PublishedApi annotation to make inline helpers less visible

mkevins committed:
Add test for access token provider helper

mkevins commented:
Thank you for testing this Renan! This is ready for another pass.

RenanLukas reviewed (approved):
Thanks for the changes, @mkevins ! The PR works as expected.

LGTM :shipit:

mkevins committed:
Merge branch 'trunk' into try/local-migration-content-provider--refactor-login-resolver-as-orchestrator

mkevins committed:
Temporarily disable login tests

These test should be refactored after the providers are unified, and the
login resolver is refactored as an orchestrator for the data migration.

mkevins committed:
Fix lint issues after merge conflict resolution

ovitrif commented:
Looking good ❤️ , nice adaptation of the self-hosted auto-login support that was recently added 🚀

mkevins committed:
Empty commit.

This is to hopefully trigger a state update on GitHub, which seems to be
having issues at the moment. A PR is showing a stale HEAD, while the
branch is showing up-to-date commits.

RenanLukas reviewed (approved):
Tested again after merge. Works as expected :shipit:

mkevins commented:
Thank you both for testing and reviewing!
